url89849c8c5b_timmolter_XChart_tgz-pJ8-org_knowm_xchart_standalone_Example1J8/src/org/knowm/xchart/CSVExporter.java

```diff
@@ -0,0 +1,200 @@
+/**
+ * Copyright 2015-2016 Knowm Inc. (http://knowm.org) and contributors.
+ * Copyright 2011-2015 Xeiam LLC (http://xeiam.com) and contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.knowm.xchart;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.io.Writer;
+import java.util.Collection;
+import java.util.Iterator;
+
+/**
+ * This class is used to export Chart data to a folder containing one or more CSV files. The parent folder's name is the title of the chart. Each
+ * series becomes a CSV file in the folder. The series' name becomes the CSV files' name.
+ *
+ * @author timmolter
+ */
+public class CSVExporter {
+
+  /**
+   * Export all XYChart series as rows in separate CSV files.
+   *
+   * @param chart
+   * @param path2Dir
+   */
+  public static void writeCSVRows(XYChart chart, String path2Dir) {
+
+    for (XYSeries xySeries : chart.getSeriesMap().values()) {
+      writeCSVRows(xySeries, path2Dir);
+    }
+  }
+
+  /**
+   * Export a XYChart series into rows in a CSV file.
+   *
+   * @param series
+   * @param path2Dir - ex. "./path/to/directory/" *make sure you have the '/' on the end
+   */
+  public static void writeCSVRows(XYSeries series, String path2Dir) {
+
+    File newFile = new File(path2Dir + series.getName() + ".csv");
+    Writer out = null;
+    try {
+
+      out = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(newFile), "UTF8"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out*. The resource is referenced at lines 61, 63, 65, 68, 73, 74, 76. The resource is closed at line 77. There are other execution paths that do not contain closure statements, for example, when Writer.flush throws an exception. Ensure *out* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).